ffi/src/lib.rs

```diff
@@ -1,23 +1,136 @@
 //! Structures, macros related to FFI and generation of FFI bindings.
+//! [Non-robust types](https://anssi-fr.github.io/rust-guide/07_ffi.html#non-robust-types-references-function-pointers-enums)
+//! are strictly avoided in the FFI API
+//!
+//! # Conversions:
+//! owned type -> opaque pointer
+//! reference -> raw pointer
+//!
+//! enum -> int
+//! bool -> u8
+//!
+//! # Conversions (WebAssembly):
+//! u8, u16 -> u32
+//! i8, i16 -> i32
+
+use core::marker::PhantomData;
 
 pub use iroha_ffi_derive::*;
 
-// NOTE: Using `u32` to be compatible with WebAssembly.
-// Otherwise `u8` should be sufficient
-/// Type of the handle id
-pub type HandleId = u32;
+pub mod handle;
+pub mod opaque;
+mod primitives;
+pub mod slice;
+
+/// Represents handle in an FFI context
+///
+/// # Safety
+///
+/// If two structures implement the same id, this may result in void pointer being casted to the wrong type
+pub unsafe trait Handle {
+    /// Unique identifier of the handle. Most commonly, it is
+    /// used to facilitate generic monomorphization over FFI
+    const ID: handle::Id;
+}
+
+/// Robust type which conforms to C ABI and can be safely shared across FFI boundaries. This does
+/// not ensure the ABI of the referent for pointers in which case the pointer should be made opaque
+///
+/// # Safety
+///
+/// Type implementing the trait must be a robust type with a guaranteed C ABI
+pub unsafe trait ReprC {}
+
+pub trait AsReprC {
+    /// Robust C ABI compliant representation of [`Self`]
+    type Item: ReprC;
 
-/// FFI compatible tuple with 2 elements
-#[repr(C)]
-#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-pub struct Pair<K, V>(pub K, pub V);
+    fn as_repr_c(&self) -> Self::Item;
+}
+
+pub trait FromReprC {
+    type Item: ReprC;
+
+    fn from_repr_c(source: Self::Item) -> Self;
+}
+
+impl<T: ReprC> AsReprC for T where T: Copy {
+    type Item = Self;
+
+    fn as_repr_c(&self) -> Self::Item {
+        *self
+    }
+}
+
+impl<T: ReprC> FromReprC for T {
+    type Item = Self;
+
+    fn from_repr_c(source: Self::Item) -> Self {
+        source
+    }
+}
+
+/// Conversion into an FFI compatible representation that consumes the input value
+///
+// TODO: Make it an unsafe trait?
+pub trait IntoFfi: Sized {
+    type Item: AsReprC;
+
+    /// Performs the conversion from [`Self`] into [`Self::Item`]
+    fn into_ffi(self) -> Self::Item;
+}
+
+/// Conversion from an FFI compatible representation that consumes the input value
+// TODO: Make it an unsafe trait?
+pub trait TryFromFfi: Sized {
+    type Item: FromReprC;
+
+    /// Performs the fallible conversion
+    ///
+    /// # Errors
+    ///
+    /// * given pointer is null
+    /// * given id doesn't identify any known handle
+    /// * given id is not a valid enum discriminant
```

Comment:
I'd add the corresponding FfiResult enum variant for each error to create links in docs

Reply:
done